pkg/pillar/cmd/zedmanager/updatestatus.go

```diff
@@ -371,10 +371,85 @@ func doUpdate(ctx *zedmanagerContext, uuidStr string,
 	return changed
 }
 
-func doInstall(ctx *zedmanagerContext, uuidStr string,
-	config types.AppInstanceConfig, status *types.AppInstanceStatus) (bool, bool) {
+// doInstallProcessStorageEntriesWithVerifiedImage
+//  Returns: (vs, storageStatusUpdate)
+func doInstallProcessStorageEntriesWithVerifiedImage(
+	ctx *zedmanagerContext,
+	appInstUUID uuid.UUID,
+	ssPtr *types.StorageStatus) (*types.VerifyImageStatus, bool) {
+
+	changed := false
+	safename := ssPtr.Safename()
+
+	// Shortcut if image is already verified
+	vs := lookupVerifyImageStatusAny(ctx, safename, ssPtr.ImageSha256)
+	// Handle post-reboot verification in progress by allowing
+	// types.DOWNLOADED. If the verification later fails we will
+	// get a delete of the VerifyImageStatus and skip this
+	if vs == nil {
+		log.Debugf("Verifier status not found for %s sha %s",
+			safename, ssPtr.ImageSha256)
+		return nil, false
+	}
+	if vs.Expired {
```

Comment:
Close but not quite right.
Expired means that the verifier is asking zedmanager whether the verifier can delete the image.
The zedmantager then checks the config refcount and if zero tells the verifier it can go ahead and delete the image by deleting VerifyImageConfig.
Thus here you should check if the VerifyImageConfig has been deleted.

Reply:
Based on further discussing there is a race condition since a VerifyImageConfig can in theory be created after it has been deleted and before the verifier has run and deleted the VerifyImageStatus. Hence the code in the PR is fine.